<a href="https://colab.research.google.com/github/mattac22/TechLabs_DriverDrowsinessDetection/blob/maroof/DDD_Basic_CNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#INFORMATION ABOUT CNN

#TRANSFORMATION scales from interval [0, 255] to [0.0, 1.0] and then Normalization x_new = (x-mean)/std

#BATCH_SIZE = 64

#----NUMBER OF PARAMETERS CNN----

# input image has size 227*227

#conv1: 4*4*3*10 + 10 = 490 -> 4*4 filter, 3 input image channels, 10 number of filters, +10 bias parameters of each filter
#conv2: 5*5*10*15 + 15 = 3765
#conv3: 6*6*15*20 + 20 = 10820
#conv4: 3*3*20*25 + 25 = 4525

#after conv4 layer: output "image": 3*3*25 = 225

#fc1: 225*45 + 45 = 10170 -> fc1 has 45 output features
#fc2: 45*20 + 20 = 920
#fc3: 20*2 + 2 = 42

#TOTAL CNN: 30732 parameters

#----NUMBER OF PARAMETERS CNN----

#---- CURRENT RESULTS ----

# --- Epoch: 0, train loss: 0.0547, train acc: 0.9920, time: 185.72551941871643
#Epoch: 0, test loss: 0.5021, test acc: 0.8298, time: 982.3053231239319

# --- Epoch: 1, train loss: 0.0054, train acc: 0.9990, time: 1167.7591831684113
#Epoch: 1, test loss: 0.3759, test acc: 0.9193, time: 1233.2979555130005

# --- Epoch: 2, train loss: 0.0047, train acc: 0.9988, time: 1410.846270084381
#Epoch: 2, test loss: 1.0650, test acc: 0.8194, time: 1476.1876842975616

# --- Epoch: 3, train loss: 0.0022, train acc: 0.9993, time: 1649.150539636612
#Epoch: 3, test loss: 0.6330, test acc: 0.8666, time: 1713.3415627479553

# --- Epoch: 4, train loss: 0.0003, train acc: 1.0000, time: 1887.5666482448578
#Epoch: 4, test loss: 0.8888, test acc: 0.8573, time: 1951.4360206127167

# --- Epoch: 5, train loss: 0.0040, train acc: 0.9987, time: 2124.932951927185
#Epoch: 5, test loss: 0.6418, test acc: 0.8751, time: 2189.122528076172

# --- Epoch: 6, train loss: 0.0012, train acc: 0.9996, time: 2360.4124615192413
#Epoch: 6, test loss: 0.7349, test acc: 0.8451, time: 2422.3692615032196

# --- Epoch: 7, train loss: 0.0002, train acc: 1.0000, time: 2594.427927017212
#Epoch: 7, test loss: 0.5947, test acc: 0.9201, time: 2659.300325155258

#---- ----CLASSIFIED WRONG---- ----
#person E, G, S have been shifted only to the test dataset

#persons in test dataset that have been classified in a wrong way

#E    278/962  = 28.9%
#e     34/1000 = 3.4%

#G      5/499  = 1.0%
#g    108/109  = 99.1%

#S      8/487  = 1.6%
#s    457/457  = 100.0%

#k      1/538  = 0.19%

#---- CURRENT RESULTS ----

In [2]:
from torch.utils.data import Dataset, DataLoader
from torchvision.datasets import ImageFolder
from torchvision import transforms

from PIL import Image

import cv2

import torch
import torch.nn as nn
import torch.nn.functional as F
import pandas as pd
import re
import os
import numpy as np

import math

In [3]:
#set the device
if torch.cuda.is_available():
  device = torch.device("cuda:0")
  print("GPU")
else:
  device = torch.device("cpu")
  print("CPU")

GPU


In [4]:
!pip install torch
!pip install torchvision

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [5]:
#mount google drive for having access to the dataset in google colab
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [10]:
#unzip if files have not been unzipped before
!unzip -uq "/content/drive/MyDrive/Driver_Drowsiness/Driver_Drowsiness_Dataset.zip"

In [26]:
#DATAFRAME OF ALL PICTURES, TAKEOUT SPECIFIC PERSONS, 

#create a pandas dataframe out of all pictures in the dataset
#---- ---- ----
# Add image paths, labels, file names and person IDs into lists
image_path, label, file_name, person_ID = [], [], [], []

#for the person_ID - identification criteria of participant
pattern = re.compile(r'^[a-zA-Z][a-zA-Z]?')

data_dir = "Driver Drowsiness Dataset (DDD)"

for class_name in os.listdir("Driver Drowsiness Dataset (DDD)"):  #goes through all folders in DDD folder
  for path in os.listdir("Driver Drowsiness Dataset (DDD)/"+class_name):  #goes through all files in both folders and updates file path and labels
    if class_name == 'Drowsy':
      label.append(0)
    else:
      label.append(1)
    image_path.append(os.path.join(data_dir, class_name, path))
    file_name.append(path)

    person_ID.append(pattern.findall(path)[0])  #identification of participant


#dataframe with path, label, file name, person_ID of all Persons
df = pd.DataFrame()
df['images'] = image_path
df['label'] = label
df['name'] = file_name
df['person'] = person_ID

print('df')
print(df.shape)
#---- ---- ----

#split pandas dataframe with all pictures into dfTest with person E, G, D and dfWithout with all pictures except of E, G, D
#---- ---- ---- ---- ---- ----
#add Person E, D and G only to the Test Dataset
dfTest = pd.concat([df[df['person']=='e'], df[df['person']=='g'], df[df['person']=='E'], df[df['person']=='G'], df[df['person']=='D'], df[df['person']=='d']])
print('dfTest')
print(dfTest.shape)


#remove Person E, D and G from the dataset with all Persons
dfWithout = df
dfWithout = dfWithout.drop(index=(df[df['person']=='e'].index))
dfWithout = dfWithout.drop(index=(df[df['person']=='E'].index))
dfWithout = dfWithout.drop(index=(df[df['person']=='g'].index))
dfWithout = dfWithout.drop(index=(df[df['person']=='G'].index))

dfWithout = dfWithout.drop(index=(df[df['person']=='d'].index))
dfWithout = dfWithout.drop(index=(df[df['person']=='D'].index))

print('dfWithout')
print(dfWithout.shape)
#---- ---- ---- ---- ---- ----


#converting imagepath and label of takeout persons G, E, D into an nparray
#----
imagePathTakeout = dfTest['images']
labelTakeout = dfTest['label']

imagePathTO = np.array(imagePathTakeout).reshape([-1,1])
labelTO = np.array(labelTakeout).reshape([-1,1])
testTO = np.hstack((imagePathTO, labelTO))

print('testTO')
print(testTO.shape)
#----


#converting imagepath and label of all other persons into an nparray
#---- ----
imagePathWithout = dfWithout['images']
labelWithout = dfWithout['label']

image_path = np.array(imagePathWithout).reshape([-1,1])
label = np.array(labelWithout).reshape([-1,1])
comp_data = np.hstack((image_path, label))
#---- ----

df
(41793, 4)
dfTest
(3754, 4)
dfWithout
(38039, 4)
testTO
(3754, 2)


In [12]:
print(testTO)

[['Driver Drowsiness Dataset (DDD)/Non Drowsy/e0860.png' 1]
 ['Driver Drowsiness Dataset (DDD)/Non Drowsy/e0210.png' 1]
 ['Driver Drowsiness Dataset (DDD)/Non Drowsy/e0303.png' 1]
 ...
 ['Driver Drowsiness Dataset (DDD)/Non Drowsy/s0185.png' 1]
 ['Driver Drowsiness Dataset (DDD)/Non Drowsy/s0100.png' 1]
 ['Driver Drowsiness Dataset (DDD)/Non Drowsy/s0257.png' 1]]


In [13]:
#number of images of ALL persons of the dataset
df.groupby("person").size()

person
A     1411
B      315
C      335
D      179
E      962
F      415
G      499
H      508
I     1095
J      474
K      630
L      732
M      777
N     1156
O     1097
P      963
Q      562
R      204
S      487
T      933
U      420
V      653
W     1162
X     1749
Y     1112
ZA     621
ZB    1551
ZC    1346
a     1252
b      409
c      400
d     1005
e     1000
g      109
h      571
i     1045
j      717
k      538
l      381
m      473
n      957
o      671
p      190
q      521
r      522
s      457
u      510
v     1002
w      493
x     1143
y     1500
za    1054
zb    1237
zc    1288
dtype: int64

In [27]:
#CREATE TRAIN DATASET AND TEST DATASET

from sklearn.model_selection import train_test_split


#nparray with all other persons (except of E, G, S) gets split into train and test dataset
trainW, testW = train_test_split(comp_data, test_size=0.2, random_state=1234)
print('trainW')
print(trainW.shape)

print('testW')
print(testW.shape)


#train dataset
train = trainW


#test dataset is a combination of the takeout of person E, G, S and the part that was split to test via train_test_split
test = np.concatenate((testTO, testW))


print('test')
print(test.shape)
print(train)

trainW
(30431, 2)
testW
(7608, 2)
test
(11362, 2)
[['Driver Drowsiness Dataset (DDD)/Non Drowsy/i0470.png' 1]
 ['Driver Drowsiness Dataset (DDD)/Non Drowsy/b0018.png' 1]
 ['Driver Drowsiness Dataset (DDD)/Non Drowsy/n0921.png' 1]
 ...
 ['Driver Drowsiness Dataset (DDD)/Drowsy/N1081.png' 0]
 ['Driver Drowsiness Dataset (DDD)/Drowsy/I0113.png' 0]
 ['Driver Drowsiness Dataset (DDD)/Drowsy/Y0264.png' 0]]


In [15]:
#basic methods of Dataset class need to be overwritten
class DDDdataset(Dataset):
  def __init__(self, image_paths, image_labels, transform):
    super().__init__()
    self.paths = image_paths
    self.labels = image_labels
    self.len = len(self.paths)
    self.transform = transform
  
  def __len__(self): return self.len
  
  def __getitem__(self, index):
    path = self.paths[index]
    label = self.labels[index]
    target_tensor=torch.from_numpy(np.array(label,dtype=np.int16))
    #image = Image.open(path).convert('RGB')
    image = cv2.imread(path,cv2.IMREAD_UNCHANGED)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    image = self.transform(image)
    return (image, target_tensor, path)

In [16]:
#TRANSFORMATION, TRAIN DATASET AND LOADER FOR THE CALCULATION OF MEAN AND STD FOR EVERY CHANNEL

#transformation for the calculation of mean and std for the train dataset
transformationCalc = transforms.Compose([
    #ToTensor() includes an automatic scaling from the interval (0,255) to (0.0,1.0)
    transforms.ToTensor()
])

#train dataset and loader for the calculation of mean and std for the train dataset
trainDatasetCalc = DDDdataset(train[:,0], train[:,-1], transformationCalc)

loaderCalc = DataLoader(trainDatasetCalc, batch_size=128, shuffle=True)

In [17]:
#CALCULATION MEAN AND STD FOR EVERY CHANNEL

numberOfPixelsPerImage = 227*227
noOfImages = train.shape[0]


#lists to collect the sum of all pixel values in the batch for R, G, B
sumOfBatchesR, sumOfBatchesG, sumOfBatchesB = [], [], []
squaredSumOfBatchesR, squaredSumOfBatchesG, squaredSumOfBatchesB = [], [], []


#in calcBatch(es) with 128 pictures each perform sum operations
for calcBatch in loaderCalc:
  #calcBatch[0] with dimension (128, 3, 227, 227)
  currentBatch = calcBatch[0]
  #currentBatch = currentBatch.float()


  rgbSumOfBatch = currentBatch.sum(axis=3) #reduce column of one picture to one column (128, 3, 227)
  rgbSumOfBatch = rgbSumOfBatch.sum(axis=2) #reduce row of one picture to one row (128, 3)
  rgbSumOfBatch = rgbSumOfBatch.sum(axis=0) #combine all pictures of the batch to 3 values for rgb (3)

  #sum of the batch for R, G, B is stored for final calculation
  sumOfBatchesR.append(rgbSumOfBatch[0])
  sumOfBatchesG.append(rgbSumOfBatch[1])
  sumOfBatchesB.append(rgbSumOfBatch[2])


  rgbSquaredSumOfBatch = ((currentBatch)**2).sum(axis=3) #reduce column of one picture to one column (128, 227, 3)
  rgbSquaredSumOfBatch = rgbSquaredSumOfBatch.sum(axis=2) #reduce row of one picture to one row (128, 3)
  rgbSquaredSumOfBatch = rgbSquaredSumOfBatch.sum(axis=0) #combine all pictures of the batch to 3 values for rgb (3)

  #squared sum of the batch for R, G, B is stored for final calculation
  squaredSumOfBatchesR.append(rgbSquaredSumOfBatch[0])
  squaredSumOfBatchesG.append(rgbSquaredSumOfBatch[1])
  squaredSumOfBatchesB.append(rgbSquaredSumOfBatch[2])


#final sum for R, G, B
sumOfBatchesR = np.array(sumOfBatchesR).sum()
sumOfBatchesG = np.array(sumOfBatchesG).sum()
sumOfBatchesB = np.array(sumOfBatchesB).sum()

#calculation of the mean for R, G, B
meanR = sumOfBatchesR/(numberOfPixelsPerImage*noOfImages)
meanG = sumOfBatchesG/(numberOfPixelsPerImage*noOfImages)
meanB = sumOfBatchesB/(numberOfPixelsPerImage*noOfImages)

#final squared sum for R, G, B
squaredSumOfBatchesR = np.array(squaredSumOfBatchesR).sum()
squaredSumOfBatchesG = np.array(squaredSumOfBatchesG).sum()
squaredSumOfBatchesB = np.array(squaredSumOfBatchesB).sum()

#final variance
varianceR = (squaredSumOfBatchesR - ((sumOfBatchesR)**2)/(numberOfPixelsPerImage*noOfImages))/(numberOfPixelsPerImage*noOfImages)
varianceG = (squaredSumOfBatchesG - ((sumOfBatchesG)**2)/(numberOfPixelsPerImage*noOfImages))/(numberOfPixelsPerImage*noOfImages)
varianceB = (squaredSumOfBatchesB - ((sumOfBatchesB)**2)/(numberOfPixelsPerImage*noOfImages))/(numberOfPixelsPerImage*noOfImages)

#final std of R, G, B
stdR = math.sqrt(varianceR)
stdG = math.sqrt(varianceG)
stdB = math.sqrt(varianceB)

print(meanR)
print(meanG)
print(meanB)
print('---')
print(stdR)
print(stdG)
print(stdB)
  

KeyboardInterrupt: ignored

In [18]:
#use this cell to avoid duration of the calculation of mean and std
meanR = 0.5078696757850544
meanG = 0.383553314333379
meanB = 0.3425597905213773

stdR = 0.2523046539381231
stdG = 0.22919015141761262
stdB =0.21377369693134016

In [19]:
#TRANSFORMATION, TRAIN DATASET AND LOADER FOR MODEL TRAINING

#transformation for model training and testing
transformation = transforms.Compose([
    #ToTensor() includes an automatic scaling from the interval [0,255] to [0.0,1.0]
    transforms.ToTensor(),
    #Normalization x_new = (x - mean)/std
    transforms.Normalize(mean=[meanR, meanG, meanB], std=[stdR, stdG, stdB])
])


#datasets and dataloaders for model training and testing
trainDataset = DDDdataset(train[:,0], train[:,-1], transformation)
testDataset = DDDdataset(test[:,0], test[:,-1], transformation)

trainDataLoader = DataLoader(trainDataset, batch_size=64, shuffle=True)
testDataLoader = DataLoader(testDataset, batch_size=64, shuffle=True)

In [20]:
#INFORMATION ABOUT CNN

#transformation scales from interval [0, 255] to [0.0, 1.0] and then Normalization

#batch_size = 64

#----NUMBER OF PARAMETERS CNN below----

# input image has size 227*227

#conv1: 4*4*3*10 + 10 = 490 -> 4*4 filter, 3 input image channels, 10 number of filters, +10 bias parameters of each filter
#conv2: 5*5*10*15 + 15 = 3765
#conv3: 6*6*15*20 + 20 = 10820
#conv4: 3*3*20*25 + 25 = 4525

#after conv4 layer: output "image": 3*3*25 = 225

#fc1: 225*45 + 45 = 10170 -> fc1 has 45 output features
#fc2: 45*20 + 20 = 920
#fc3: 20*2 + 2 = 42

#TOTAL CNN: 30732 parameters

#----NUMBER OF PARAMETERS CNN below----

In [21]:
#LAYERS OF THE MODEL

class CNN(nn.Module):
  def __init__(self):
    super().__init__()

    self.conv1 = nn.Conv2d(in_channels=3, out_channels=10, kernel_size=(4,4), stride=1, padding=1)
    self.bn1 = nn.BatchNorm2d(10)
    self.conv2 = nn.Conv2d(in_channels=10, out_channels=15, kernel_size=(5,5), stride=2, padding=1)
    self.bn2 = nn.BatchNorm2d(15)
    self.conv3 = nn.Conv2d(in_channels=15, out_channels=20, kernel_size=(6,6), stride=2, padding=1)
    self.bn3 = nn.BatchNorm2d(20)
    self.conv4 = nn.Conv2d(in_channels=20, out_channels=25, kernel_size=(3,3), stride=1, padding=1)
    self.bn4 = nn.BatchNorm2d(25)

    self.fc1 = nn.Linear(in_features=25*3*3, out_features=45)
    self.bn5 = nn.BatchNorm1d(45)
    self.fc2 = nn.Linear(in_features=45, out_features=20)
    self.bn6 = nn.BatchNorm1d(20)
    self.fc3 = nn.Linear(in_features=20, out_features=2)

  def forward(self, X):

    X = F.relu(self.bn1(self.conv1(X)))
    X = F.max_pool2d(X, 2)
    X = F.relu(self.bn2(self.conv2(X)))
    X = F.max_pool2d(X, 2)
    X = F.relu(self.bn3(self.conv3(X)))
    X = F.max_pool2d(X, 3, stride=2)
    X = F.relu(self.bn4(self.conv4(X)))
    X = F.max_pool2d(X, 2)

    X = X.view(X.shape[0], -1)
    X = F.relu(self.bn5(self.fc1(X)))
    X = F.relu(self.bn6(self.fc2(X)))
    X = self.fc3(X)

    return X


In [28]:
#MODEL TRAINING AND TESTING

import time

#basic settings
model = CNN().to(device)
number_epochs = 8
loss_func = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
losses = []
accuracies = []


#for the analysis of the wrong classifications
pathsOfWrongClassifications = []
rowsOfWrongClassifications = []
correctLabel = []
epochOfWrongClassification = []


start = time.time()
#model training and testing process
for epoch in range(number_epochs):


  epoch_loss = 0
  epoch_accuracy = 0
  
  #-------model training-------
  for (X, y, path) in trainDataLoader:
    X = X.to(device) #torch.as_tensor(X)
    y = y.to(device) #torch.as_tensor(y)
    predictions = model(X)
    loss = loss_func(predictions, y.long())

    #adaption of the neural network
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    #
    accuracy = ((predictions.argmax(dim=1) == y).float().mean())
    epoch_accuracy += accuracy
    epoch_loss += loss
  
  #total epoch_accuracy and epoch_loss calculation
  epoch_accuracy = epoch_accuracy/len(trainDataLoader)
  accuracies.append(epoch_accuracy)
  epoch_loss = epoch_loss / len(trainDataLoader)
  losses.append(epoch_loss)

  print("\n --- Epoch: {}, train loss: {:.4f}, train acc: {:.4f}, time: {}".format(epoch, epoch_loss, epoch_accuracy, time.time() - start))
  #-------model training-------
  
  
  #---model testing---
  model.eval() #sets the model in evaluation mode
  
  #with disabled gradient calculation
  with torch.no_grad():
    test_epoch_loss = 0
    test_epoch_accuracy = 0

    for test_X, test_y, path in testDataLoader:

      test_X = test_X.to(device)
      test_y = test_y.to(device)

      test_preds = model(test_X)
      
      test_loss = loss_func(test_preds, test_y.long())

      test_epoch_loss += test_loss            
      test_accuracy = ((test_preds.argmax(dim=1) == test_y).float().mean())
      test_epoch_accuracy += test_accuracy

      #if test prediction is wrong save the path of the image that has been classified wrong
      for index, row in enumerate(test_preds):
        if row.argmax(dim=0) != test_y[index]:
          pathsOfWrongClassifications.append(path[index])
          if rowsOfWrongClassifications == []:
            row = row.cpu()
            rowsOfWrongClassifications = row
          else:
            row = row.cpu()
            rowsOfWrongClassifications = np.vstack([rowsOfWrongClassifications, row])
          correctLabel.append(test_y[index].item())
          epochOfWrongClassification.append(epoch)

    #total test_epoch_accuracy and test_epoch_loss calculation
    test_epoch_accuracy = test_epoch_accuracy/len(testDataLoader)
    test_epoch_loss = test_epoch_loss / len(testDataLoader)

    print("Epoch: {}, test loss: {:.4f}, test acc: {:.4f}, time: {}\n".format(epoch, test_epoch_loss, test_epoch_accuracy, time.time() - start))
  #---model testing---

  model.train(mode=True) #sets the model back in training mode



 --- Epoch: 0, train loss: 0.0613, train acc: 0.9911, time: 99.21207880973816


<ipython-input-28-96f23ff36f87>:81: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  if rowsOfWrongClassifications == []:


Epoch: 0, test loss: 0.2987, test acc: 0.8995, time: 137.28974080085754


 --- Epoch: 1, train loss: 0.0019, train acc: 0.9999, time: 235.76221013069153
Epoch: 1, test loss: 0.1005, test acc: 0.9717, time: 273.4768898487091


 --- Epoch: 2, train loss: 0.0007, train acc: 1.0000, time: 371.6678569316864
Epoch: 2, test loss: 0.1812, test acc: 0.9248, time: 409.57978773117065


 --- Epoch: 3, train loss: 0.0106, train acc: 0.9968, time: 507.6580882072449
Epoch: 3, test loss: 0.5380, test acc: 0.8958, time: 545.365490436554


 --- Epoch: 4, train loss: 0.0028, train acc: 0.9993, time: 644.4032781124115
Epoch: 4, test loss: 0.3826, test acc: 0.8615, time: 682.2523202896118


 --- Epoch: 5, train loss: 0.0004, train acc: 0.9999, time: 780.6437330245972
Epoch: 5, test loss: 0.1733, test acc: 0.9231, time: 818.3183073997498


 --- Epoch: 6, train loss: 0.0002, train acc: 1.0000, time: 916.3153054714203
Epoch: 6, test loss: 0.2157, test acc: 0.9257, time: 953.5379910469055


 --- Epoch: 7, trai

In [30]:
#DATAFRAME OF ALL PERSONS THAT HAVE BEEN CLASSIFIED WRONG IN ALL EPOCHS

#extract the person ID out of the stored image paths
personID = []
for pathStr in pathsOfWrongClassifications:
  personID.append(pathStr[pathStr.rfind('/')+1:-8])


#create a dataframe from all wrong classified images in the test dataset
dfWrong = pd.DataFrame()
dfWrong['testEpochNo'] = epochOfWrongClassification
dfWrong['personID'] = personID
dfWrong['imagepath'] = pathsOfWrongClassifications
dfWrong['correctLabel'] = correctLabel
dfWrong['outputValueLabel0'] = rowsOfWrongClassifications[:,0]
dfWrong['outputValueLabel1'] = rowsOfWrongClassifications[:,1]

print(dfWrong)

      testEpochNo personID                                          imagepath  \
0               0        e  Driver Drowsiness Dataset (DDD)/Non Drowsy/e01...   
1               0        e  Driver Drowsiness Dataset (DDD)/Non Drowsy/e05...   
2               0        e  Driver Drowsiness Dataset (DDD)/Non Drowsy/e04...   
3               0        e  Driver Drowsiness Dataset (DDD)/Non Drowsy/e05...   
4               0        e  Driver Drowsiness Dataset (DDD)/Non Drowsy/e03...   
...           ...      ...                                                ...   
7030            7        e  Driver Drowsiness Dataset (DDD)/Non Drowsy/e01...   
7031            7        g  Driver Drowsiness Dataset (DDD)/Non Drowsy/g01...   
7032            7        e  Driver Drowsiness Dataset (DDD)/Non Drowsy/e03...   
7033            7        e  Driver Drowsiness Dataset (DDD)/Non Drowsy/e10...   
7034            7        e  Driver Drowsiness Dataset (DDD)/Non Drowsy/e05...   

      correctLabel  outputV

In [31]:
#get an overview of the persons that are likely to be classified in a wrong way (from last epoch)
dfOfEpoch = dfWrong[dfWrong['testEpochNo']==7]
dfOfEpoch.groupby('personID').size()

personID
E      9
G      7
e    303
g     66
dtype: int64

Results of older different trials:


I)

2 epochs

persons that have been shifted totally to the training dataset:
e, E, g, G

wrong classifications:

E: 5/962 = 0.5%

G: 48/499 = 9.6%

e: 1000/1000 = 100%

g: 97/109 = 89.0%

s: 18/457 = 3.9%

S: 0/487 = 0.0%

all wrong classifications:

personID

E       5

G      48

b       7

c       1

e    1000

g      97

n      10

o       2

r       2

s      18

u       8

w       5

x       1

y       3

II)

3 epochs

persons that have been shifted totally to the training dataset:
e, E, g, G, s, S

wrong classifications:

E: 888/962 = 92.3%

G: 0/499 = 0.0%

e: 854/1000 = 85.4%

g: 11/109 = 10.1%

s: 402/457 = 88.0%

S: 419/487 = 86.0%


all wrong classifications

personID

E    888

S    419

e    854

g     11

s    402

